File: 383-ransom-note/ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        HashMap<Character, Integer> ransomNoteCounts = new HashMap<>();
        HashMap<Character, Integer> magazineCounts = new HashMap<>();
        for (char c : ransomNote.toCharArray()) {
            ransomNoteCounts.put(c, ransomNoteCounts.getOrDefault(c, 0) + 1);
        }
        for (char c : magazine.toCharArray()) {
            magazineCounts.put(c, magazineCounts.getOrDefault(c, 0) + 1);
        }
        for (Map.Entry<Character, Integer> entry : ransomNoteCounts.entrySet()) {
            char charNeeded = entry.getKey();
            int countNeeded = entry.getValue();
            int countAvailable = magazineCounts.getOrDefault(charNeeded, 0);
            if (countAvailable < countNeeded) {
                return false;
            }
        }
        return true;
    }
}

File: 383-ransom-note/ransom-note.cpp
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<char,int> mpp1;
        unordered_map<char,int> mpp2;
        for(char& c:ransomNote){
            mpp1[c]++;
        }
        for(char& c:magazine){
            mpp2[c]++;
        }
        for(auto const& [key,val]: mpp1){
            if(mpp2[key]<val){
                return false;
            }
        }
        return true;
    }
};